File: sdk/nodejs/src/private-beta.ts
import fetch, { Response } from 'node-fetch';
import { SatGateAuthError, SatGateError } from './errors';

const BETA_ACCESS_MESSAGE =
  'This API namespace requires private beta access. Visit cloud.satgate.io/docs to request access.';

export interface SatGateOptions {
  apiKey?: string;
  baseUrl?: string;
  tenant?: string;
  timeout?: number;
}

export type JsonValue = string | number | boolean | null | JsonObject | JsonValue[];
export type JsonObject = { [key: string]: JsonValue };

export interface CapabilityRequest {
  [key: string]: unknown;
  task: string;
  agent: string;
  allow: string[];
  budgetUsd?: number;
  budget_usd?: number;
  expiresIn?: string;
  expires_in?: string;
}

export interface PayRequest {
  [key: string]: unknown;
  upstream: string;
  capability: unknown;
  maxUsd?: number;
  max_usd?: number;
}

export type SatGateReceipt = JsonObject & {
  decision?: string;
  evidencePackId?: string;
  evidence_pack_id?: string;
};

export class SatGate {
  private readonly apiKey?: string;
  private readonly baseUrl: string;
  private readonly tenant?: string;
  private readonly timeout: number;

  constructor(options: SatGateOptions = {}) {
    this.apiKey = options.apiKey;
    this.baseUrl = (options.baseUrl ?? 'https://api.satgate.io').replace(/\/$/, '');
    this.tenant = options.tenant;
    this.timeout = options.timeout ?? 30000;
  }

  async issue(request: CapabilityRequest): Promise<SatGateReceipt> {
    try {
      return await this.post<SatGateReceipt>('/v1/issue', this.toApiPayload(request) as JsonObject);
    } catch (err) {
      if (err instanceof SatGateAuthError && err.statusCode === 404) {
        // Compatibility fallback for early private-beta clients and gateways.
        return this.post<SatGateReceipt>('/v1/capabilities', this.toApiPayload(request) as JsonObject);
      }
      throw err;
    }
  }

  async pay(request: PayRequest): Promise<SatGateReceipt> {
    return this.post<SatGateReceipt>('/v1/pay', this.toApiPayload(request) as JsonObject);
  }

  async verify(receipt: unknown): Promise<SatGateReceipt> {
    return this.post<SatGateReceipt>('/v1/verify', { receipt: this.toApiPayload(receipt) });
  }

  private async post<T>(path: string, body: JsonObject): Promise<T> {
    if (!this.apiKey) {
      throw new SatGateAuthError(BETA_ACCESS_MESSAGE, 401);
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);
    const headers: Record<string, string> = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
    };
    if (this.tenant) {
      headers['X-SatGate-Tenant'] = this.tenant;
    }

    let response: Response;
    try {
      response = await fetch(`${this.baseUrl}${path}`, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });
    } catch (err: unknown) {
      clearTimeout(timeoutId);
      throw new SatGateError(`Request failed: ${err}`);
    }
    clearTimeout(timeoutId);

    if ([401, 403, 404, 503].includes(response.status)) {
      throw new SatGateAuthError(BETA_ACCESS_MESSAGE, response.status);
    }

    if (!response.ok) {
      const errBody = await response.json().catch(() => ({})) as Record<string, unknown>;
      throw new SatGateError(
        (errBody.message as string) || (errBody.error as string) || `Request failed with status ${response.status}`,
        response.status
      );
    }

    if (response.status === 204) {
      return {} as T;
    }
    return response.json() as Promise<T>;
  }

  private toApiPayload(value: unknown): JsonValue {
    if (Array.isArray(value)) {
      return value.map((item) => this.toApiPayload(item));
    }
    if (value && typeof value === 'object') {
      const output: JsonObject = {};
      for (const [key, val] of Object.entries(value as Record<string, unknown>)) {
        output[this.toSnakeCase(key)] = this.toApiPayload(val);
      }
      return output;
    }
    return value as JsonValue;
  }

  private toSnakeCase(key: string): string {
    return key.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);
  }
}

File: sdk/python/satgate/private_beta.py
"""Private-beta issue/pay/verify facade for SatGate Cloud."""

from types import SimpleNamespace
from typing import Any, Dict, Optional

import requests

from .exceptions import SatGateAuthError, SatGateError

BETA_ACCESS_MESSAGE = (
    "This API namespace requires private beta access. "
    "Visit cloud.satgate.io/docs to request access."
)


class AttrDict(SimpleNamespace):
    """Small response wrapper so examples can use receipt.foo attributes."""

    @classmethod
    def from_value(cls, value: Any) -> Any:
        if isinstance(value, dict):
            return cls(**{key: cls.from_value(val) for key, val in value.items()})
        if isinstance(value, list):
            return [cls.from_value(item) for item in value]
        return value

    def to_dict(self) -> Dict[str, Any]:
        result: Dict[str, Any] = {}
        for key, value in self.__dict__.items():
            if isinstance(value, AttrDict):
                result[key] = value.to_dict()
            elif isinstance(value, list):
                result[key] = [item.to_dict() if isinstance(item, AttrDict) else item for item in value]
            else:
                result[key] = value
        return result


class SatGate:
    """SatGate Cloud issue/pay/verify client.

    The public package installs today. The issue/pay/verify API namespace is
    private beta; unauthenticated or non-beta credentials raise SatGateAuthError
    with a docs CTA instead of a stack trace or mocked receipt.
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = "https://api.satgate.io",
        tenant: Optional[str] = None,
        timeout: int = 30,
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.tenant = tenant
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Content-Type": "application/json"})
        if api_key:
            self.session.headers.update({"Authorization": f"Bearer {api_key}"})
        if tenant:
            self.session.headers.update({"X-SatGate-Tenant": tenant})

    def issue(self, **payload: Any) -> AttrDict:
        """Issue a scoped capability for an agent task."""
        return self._post("/v1/capabilities", payload)

    def pay(self, **payload: Any) -> AttrDict:
        """Route a value-bearing call under a scoped capability."""
        return self._post("/v1/pay", self._serialize(payload))

    def verify(self, receipt: Any) -> AttrDict:
        """Verify a SatGate receipt and return proof metadata."""
        payload = receipt.to_dict() if hasattr(receipt, "to_dict") else receipt
        return self._post("/v1/verify", {"receipt": self._serialize(payload)})

    def _post(self, path: str, payload: Dict[str, Any]) -> AttrDict:
        if not self.api_key:
            raise SatGateAuthError(BETA_ACCESS_MESSAGE, status_code=401)

        try:
            response = self.session.post(
                f"{self.base_url}{path}",
                json=payload,
                timeout=self.timeout,
            )
        except requests.RequestException as exc:
            raise SatGateError(f"Request failed: {exc}") from exc

        if response.status_code in {401, 403, 404, 503}:
            raise SatGateAuthError(BETA_ACCESS_MESSAGE, status_code=response.status_code)

        if response.status_code >= 400:
            try:
                body = response.json()
                message = body.get("message") or body.get("error") or response.text
            except Exception:
                message = response.text
            raise SatGateError(f"API error {response.status_code}: {message}")

        if not response.content:
            return AttrDict()
        return AttrDict.from_value(response.json())

    def _serialize(self, value: Any) -> Any:
        if isinstance(value, AttrDict):
            return value.to_dict()
        if isinstance(value, dict):
            return {key: self._serialize(val) for key, val in value.items()}
        if isinstance(value, list):
            return [self._serialize(item) for item in value]
        return value
